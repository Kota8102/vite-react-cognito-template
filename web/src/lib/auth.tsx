import { useEffect } from "react";
import type { ReactNode } from "react";
import { useAuth } from "react-oidc-context";
import { useNavigate } from "react-router-dom";
import { LoadingSpinner } from "../components/loading";
import { cognitoLogoutConfig } from "../config/auth";
import { paths } from "../config/paths";

/**
 * Cognitoへのログインリダイレクト処理
 */
export const loginRedirect = (auth: ReturnType<typeof useAuth>) => {
	auth.signinRedirect({
		extraQueryParams: {
			ui_locales: "ja",
			language: "ja",
		},
	});
};

/**
 * ログアウト処理
 * Cognitoのログアウトエンドポイントにリダイレクトしてセッションを終了します
 * ログアウト後は/loginページに遷移します
 */
export const signOut = (auth: ReturnType<typeof useAuth>) => {
	const { clientId } = cognitoLogoutConfig;
	const logoutUri = `${window.location.origin}${paths.login.path}`;

	auth.signoutRedirect({
		extraQueryParams: {
			client_id: clientId,
			logout_uri: logoutUri,
		},
	});
};

/**
 * 認証が必要なコンポーネントをラップするためのコンポーネント
 * 認証されていない場合は、/loginにリダイレクトします
 */
export const RequireAuth = ({ children }: { children: ReactNode }) => {
	const auth = useAuth();
	const navigate = useNavigate();
	const { isLoading, isAuthenticated, error } = auth;

	useEffect(() => {
		if (!isLoading && !isAuthenticated && !error) {
			navigate(paths.login.path);
		}
	}, [isLoading, isAuthenticated, error, navigate]);

	if (isLoading) {
		return <LoadingSpinner />;
	}

	if (error) {
		return (
			<div className="min-h-screen flex flex-col items-center justify-center">
				<div className="text-2xl text-red-500 mb-4">
					認証エラーが発生しました
				</div>
				<div className="text-lg">{error.message}</div>
			</div>
		);
	}

	if (!isAuthenticated) {
		return <LoadingSpinner />;
	}

	return <>{children}</>;
};
