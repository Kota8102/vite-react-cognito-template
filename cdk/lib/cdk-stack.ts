import * as cdk from "aws-cdk-lib";
import type { Construct } from "constructs";
import { Auth } from "./construct/auth";

export class CdkStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Cognitoの設定
    new Auth(this, "Auth", {
      stackName: id,
    });
  }
}
