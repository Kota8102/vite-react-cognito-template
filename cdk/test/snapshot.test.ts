import { App } from "aws-cdk-lib";
import { describe, expect, test } from "vitest";
import { CdkStack } from "../lib/cdk-stack";

describe("CDK Snapshot Tests", () => {
  test("Matches infrastructure snapshot", () => {
    const app = new App();
    const stack = new CdkStack(app, "TestStack");
    const template = JSON.stringify(
      app.synth().getStackArtifact(stack.artifactId).template,
      null,
      2
    );
    expect(template).toMatchSnapshot();
  });
});
