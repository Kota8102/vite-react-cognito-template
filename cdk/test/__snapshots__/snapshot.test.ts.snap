// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CDK Snapshot Tests > Matches infrastructure snapshot 1`] = `
"{
  "Resources": {
    "AuthUserPool8115E87F": {
      "Type": "AWS::Cognito::UserPool",
      "Properties": {
        "AccountRecoverySetting": {
          "RecoveryMechanisms": [
            {
              "Name": "verified_email",
              "Priority": 1
            }
          ]
        },
        "AdminCreateUserConfig": {
          "AllowAdminCreateUserOnly": false
        },
        "AutoVerifiedAttributes": [
          "email"
        ],
        "EmailVerificationMessage": "The verification code to your new account is {####}",
        "EmailVerificationSubject": "Verify your new account",
        "Policies": {
          "PasswordPolicy": {
            "MinimumLength": 8,
            "RequireLowercase": true,
            "RequireNumbers": true,
            "RequireSymbols": true,
            "RequireUppercase": true
          }
        },
        "Schema": [
          {
            "Mutable": true,
            "Name": "email",
            "Required": true
          }
        ],
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "UserPoolName": "TestStack-user-pool",
        "UsernameAttributes": [
          "email"
        ],
        "VerificationMessageTemplate": {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "The verification code to your new account is {####}",
          "EmailSubject": "Verify your new account",
          "SmsMessage": "The verification code to your new account is {####}"
        }
      },
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "AuthUserPoolClient0AA456E4": {
      "Type": "AWS::Cognito::UserPoolClient",
      "Properties": {
        "AllowedOAuthFlows": [
          "implicit",
          "code"
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": [
          "profile",
          "phone",
          "email",
          "openid",
          "aws.cognito.signin.user.admin"
        ],
        "CallbackURLs": [
          "https://example.com"
        ],
        "ExplicitAuthFlows": [
          "ALLOW_USER_PASSWORD_AUTH",
          "ALLOW_ADMIN_USER_PASSWORD_AUTH",
          "ALLOW_REFRESH_TOKEN_AUTH"
        ],
        "SupportedIdentityProviders": [
          "COGNITO"
        ],
        "UserPoolId": {
          "Ref": "AuthUserPool8115E87F"
        }
      }
    }
  },
  "Outputs": {
    "AuthUserPoolIdC0605E59": {
      "Value": {
        "Ref": "AuthUserPool8115E87F"
      }
    },
    "AuthUserPoolClientId8216BF9A": {
      "Value": {
        "Ref": "AuthUserPoolClient0AA456E4"
      }
    }
  },
  "Parameters": {
    "BootstrapVersion": {
      "Type": "AWS::SSM::Parameter::Value<String>",
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]"
    }
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5"
                  ],
                  {
                    "Ref": "BootstrapVersion"
                  }
                ]
              }
            ]
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI."
        }
      ]
    }
  }
}"
`;
